File: ruler/src/main/java/hsicen/ruler/InnerRulers/CenterHorizontalRuler.java
package hsicen.ruler.InnerRulers;

import android.content.Context;
import android.graphics.Canvas;

import hsicen.ruler.BooheeRuler;
import hsicen.ruler.RulerStringUtil;


/**
 * 作者：hsicen  5/25/21 19:27
 * 邮箱：codinghuang@163.com
 * 功能：
 * 描述：刻度在中间
 */
public class CenterHorizontalRuler extends HorizontalRuler {
  public CenterHorizontalRuler(Context context, BooheeRuler booheeRuler) {
    super(context, booheeRuler);
  }

  @Override
  protected void onDraw(Canvas canvas) {
    super.onDraw(canvas);
    drawScale(canvas);
    drawEdgeEffect(canvas);
  }

  //画刻度和字
  private void drawScale(Canvas canvas) {
    float start = (getScrollX() - mDrawOffset) / mParent.getInterval() + mParent.getMinScale();
    float end = (getScrollX() + canvas.getWidth() + mDrawOffset) / mParent.getInterval() + mParent.getMinScale();
    int height = canvas.getHeight();

    for (float i = start; i <= end; i++) {
      float locationX = (i - mParent.getMinScale()) * mParent.getInterval();

      if (i >= mParent.getMinScale() && i <= mParent.getMaxScale()) {
        if (i % mCount == 0) {
          //整数刻度  绘制线和文字
          int startY = height / 2 - mParent.getBigScaleLength() / 2;
          int endY = height / 2 + mParent.getBigScaleLength() / 2;
          canvas.drawLine(locationX, startY, locationX, endY, mBigScalePaint);
          canvas.drawText(RulerStringUtil.formatValue(i, mParent.getFactor()), locationX, height - mParent.getTextMarginHead(), mTextPaint);
        } else {
          //小数刻度  绘制线
          int startY = height / 2 - mParent.getSmallScaleLength() / 2;
          int endY = height / 2 + mParent.getSmallScaleLength() / 2;
          canvas.drawLine(locationX, startY, locationX, endY, mSmallScalePaint);
        }
      }
    }

    //画轮廓线
    //canvas.drawLine(getScrollX(), canvas.getHeight(), getScrollX() + canvas.getWidth(), canvas.getHeight(), mOutLinePaint);
  }

  //画边缘效果
  private void drawEdgeEffect(Canvas canvas) {
    if (mParent.canEdgeEffect()) {
      if (!mStartEdgeEffect.isFinished()) {
        int count = canvas.save();
        canvas.rotate(270);
        canvas.translate(-getHeight(), 0);
        if (mStartEdgeEffect.draw(canvas)) {
          postInvalidateOnAnimation();
        }
        canvas.restoreToCount(count);
      } else {
        mStartEdgeEffect.finish();
      }
      if (!mEndEdgeEffect.isFinished()) {
        int count = canvas.save();
        canvas.rotate(90);
        canvas.translate((getHeight() - mParent.getCursorHeight()), -mLength);
        if (mEndEdgeEffect.draw(canvas)) {
          postInvalidateOnAnimation();
        }
        canvas.restoreToCount(count);
      } else {
        mEndEdgeEffect.finish();
      }
    }
  }
}

File: ruler/src/main/java/hsicen/ruler/inner/CenterHorizontalRuler.kt
package hsicen.ruler.inner

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.util.AttributeSet
import com.hsicen.extension.extensions.color
import com.hsicen.extension.extensions.dp2px
import hsicen.ruler.BooheeRuler
import hsicen.ruler.R
import hsicen.ruler.RulerStringUtil.formatSpecial

/**
 * 作者：hsicen  5/25/21 19:27
 * 邮箱：codinghuang@163.com
 * 功能：
 * 描述：刻度在中间
 */
class CenterHorizontalRuler @JvmOverloads constructor(
  context: Context,
  booheeRuler: BooheeRuler? = null,
  attrs: AttributeSet? = null
) : HorizontalRuler(context, booheeRuler) {

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    drawScale(canvas)
    drawEdgeEffect(canvas)
  }

  private fun drawScale(canvas: Canvas) {
    val start = ((scrollX - mDrawOffset) / mParent.interval + mParent.minScale).toFloat()
    val end = ((scrollX + canvas.width + mDrawOffset) / mParent.interval + mParent.minScale).toFloat()
    val height = canvas.height
    val startY = 20.dp2px
    var toX = start

    var index = start
    while (index <= end) {
      val locationX = (index - mParent.minScale) * mParent.interval
      if (index >= mParent.minScale && index <= mParent.maxScale) {
        toX = locationX
        if (index % mCount == 0f) {
          val endY = startY + mParent.bigScaleLength
          if (index >= mParent.limitMinScale && index <= mParent.limitMaxScale) {
            mBigScalePaint.color = color(R.color.white)
            mTextPaint.color = color(R.color.white)
          } else {
            mBigScalePaint.color = color(R.color.white9)
            mTextPaint.color = color(R.color.white9)
          }

          canvas.drawLine(locationX, startY.toFloat(), locationX, endY.toFloat(), mBigScalePaint)
          canvas.drawText("${formatSpecial(index, mParent.factor)}x", locationX, (height - mParent.textMarginHead).toFloat(), mTextPaint)
        } else {
          val endY = startY + mParent.smallScaleLength
          if (index >= mParent.limitMinScale && index <= mParent.limitMaxScale) {
            mSmallScalePaint.color = color(R.color.white29)
          } else {
            mSmallScalePaint.color = color(R.color.white9)
          }
          canvas.drawLine(locationX, startY.toFloat(), locationX, endY.toFloat(), mSmallScalePaint)
        }
      }
      index++
    }

    val fromX = 0f - 10f.dp2px
    val fromY = 5f.dp2px
    val toY = fromY + 40f.dp2px
    mOutLinePaint.color = color(R.color.white4)
    mOutLinePaint.style = Paint.Style.STROKE
    canvas.drawRoundRect(fromX, fromY, toX + 10.dp2px, toY, 8f.dp2px, 8f.dp2px, mOutLinePaint)
  }

  //画边缘效果
  private fun drawEdgeEffect(canvas: Canvas) {
    if (mParent.canEdgeEffect) {
      if (!mStartEdgeEffect.isFinished) {
        val count = canvas.save()
        canvas.rotate(270f)
        canvas.translate(-height.toFloat(), 0f)
        if (mStartEdgeEffect.draw(canvas)) {
          postInvalidateOnAnimation()
        }
        canvas.restoreToCount(count)
      } else {
        mStartEdgeEffect.finish()
      }

      if (!mEndEdgeEffect.isFinished) {
        val count = canvas.save()
        canvas.rotate(90f)
        canvas.translate((height - mParent.cursorHeight).toFloat(), -mLength.toFloat())
        if (mEndEdgeEffect.draw(canvas)) {
          postInvalidateOnAnimation()
        }
        canvas.restoreToCount(count)
      } else {
        mEndEdgeEffect.finish()
      }
    }
  }
}